transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu

```diff
@@ -944,4 +957,178 @@ void dispatch_scaled_masked_softmax_backward(
+    float scale_factor,
+    cudaStream_t stream) {
+
+    // output grads is a 4d tensor with dimensions [batches, attn_heads, seq_len, seq_len]
+    const int batches = output_grads.shape[0];
+    const int attn_heads = output_grads.shape[1];
+    const int query_seq_len = output_grads.shape[2];
+    const int key_seq_len = output_grads.shape[3];
+
+    // Softmax Grad
+    TRANSFORMER_ENGINE_TYPE_SWITCH_16BIT(output_grads.dtype, softmax_type,
+        dispatch_scaled_masked_softmax_backward<softmax_type, softmax_type, float>(
+            reinterpret_cast<softmax_type*>(output_grads.dptr),
+            reinterpret_cast<softmax_type*>(output_grads.dptr),
+            reinterpret_cast<softmax_type const*>(softmax_results.dptr),
+            scale_factor,
+            query_seq_len,
+            key_seq_len,
+            batches,
+            attn_heads,
+            stream););
+}
+
+
+void scaled_masked_softmax_forward_cuda(
+    const Tensor input,
+    const Tensor mask,
+    Tensor *softmax_results,
+    float scale_factor,
+    cudaStream_t stream) {
+
+    const int batches = input.shape[0];
+    const int pad_batches = mask.shape[0];
+    const int attn_heads = input.shape[1];
+    const int query_seq_len = input.shape[2];
+    const int key_seq_len = input.shape[3];
+
+    TRANSFORMER_ENGINE_TYPE_SWITCH_16BIT(input.dtype, softmax_type,
+        dispatch_scaled_masked_softmax_forward<softmax_type, softmax_type, float>(
+            reinterpret_cast<softmax_type*>(softmax_results->dptr),
+            reinterpret_cast<const softmax_type*>(input.dptr),
+            reinterpret_cast<const uint8_t*>(mask.dptr),
+            scale_factor,
+            query_seq_len,
+            key_seq_len,
+            batches,
+            attn_heads,
+            pad_batches,
+            stream););
+}
+
+
+void scaled_masked_softmax_backward_cuda(
+    const Tensor output_grads,
+    const Tensor softmax_results,
+    float scale_factor,
+    cudaStream_t stream
+)  {
+    // output grads is a 4d tensor with dimensions [batches, attn_heads, seq_len, seq_len]
+    const int batches = output_grads.shape[0];
+    const int attn_heads = output_grads.shape[1];
+    const int query_seq_len = output_grads.shape[2];
+    const int key_seq_len = output_grads.shape[3];
+
+    // Softmax Grad
+    TRANSFORMER_ENGINE_TYPE_SWITCH_16BIT(output_grads.dtype, softmax_type,
+        dispatch_scaled_masked_softmax_backward<softmax_type, softmax_type, float>(
+            reinterpret_cast<softmax_type*>(output_grads.dptr),
+            reinterpret_cast<softmax_type*>(output_grads.dptr),
+            reinterpret_cast<softmax_type const*>(softmax_results.dptr),
+            scale_factor,
+            query_seq_len,
+            key_seq_len,
+            batches,
+            attn_heads,
+            stream););
+}
+
+
+}  // end namespace transformer_engine
+
+
+int get_batch_per_block_cuda(int query_seq_len, int key_seq_len, int batches, int attn_heads) {
```

Comment:
What is with all those `_cuda` names?

Reply:
Also, those should probably live in some anonymous namespace or something.

Reply:
Re: anonymous namespace; we need this visible for the extension

Reply:
We could remove this function from the cpp side and the extensions entirely (currently it is also exposed in the API header `softmax.h`) and move it to the python side in `softmax.py`.

Reply:
Both things are bad, but let's move it to Python side - it is better than having this random internal function in the C API. Going forward we need to make sure that this softmax API supports all shapes.